app/aboot/lk2nd-motorola.c

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include <debug.h>
+#include <libfdt.h>
+#include <malloc.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include "lk2nd-device.h"
+#include "lk2nd-motorola.h"
+#include "smem.h"
+
+#define SMEM_ID_VENDOR0          0x86
+#define SMEM_KERNEL_RESERVE      SMEM_ID_VENDOR0
+#define SMEM_KERNEL_RESERVE_SIZE 1024
+
+void lk2nd_motorola_smem_write_unit_info(const void *fdt, int offset)
+{
+	int len = 0;
+	fdt_getprop(fdt, offset, "lk2nd,motorola-unitinfo", &len);
+	if (len == 0) {
+		return;
+	}
+
+	struct mmi_unit_info *unit_info;
+	unit_info = malloc(SMEM_KERNEL_RESERVE_SIZE);
+	memset(unit_info, 0, SMEM_KERNEL_RESERVE_SIZE);
+
+	int chosen_offset = fdt_path_offset(fdt, "/chosen");
+	const uint32_t *system_rev = fdt_getprop(fdt, chosen_offset, "linux,hwrev", NULL);
+	const uint32_t *system_serial_low = fdt_getprop(fdt, chosen_offset, "linux,seriallow", NULL);
+	const uint32_t *system_serial_high = fdt_getprop(fdt, chosen_offset, "linux,serialhigh", NULL);
+	const uint32_t *powerup_reason = fdt_getprop(fdt, chosen_offset, "mmi,powerup_reason", NULL);
+
+	const char *baseband = fdt_getprop(fdt, chosen_offset, "mmi,baseband", &len);
+	if (len == -1) {
+		baseband = "";
+	}
+
+	unit_info->version = MMI_UNIT_INFO_VER;
+	unit_info->system_rev = fdt32_to_cpu(*system_rev);
+	unit_info->system_serial_low = fdt32_to_cpu(*system_serial_low);
+	unit_info->system_serial_high = fdt32_to_cpu(*system_serial_high);
```

Comment:
This will crash if the properties are missing in the device tree. Here a helper method might be useful:

```c
static inline void fdt_getprop_u32(const void *fdt, int offset, const char *name, uint32_t *val) {
	int len;
	const uint32_t *prop = fdt_getprop(fdt, chosen_offset, "linux,hwrev", &len);
	if (len == sizeof(*val))
		*val = fdt32_to_cpu(*prop);
}
```

```c
fdt_getprop_u32(fdt, chosen_offset, "linux,hwrev", &unit_info->system_rev);
```

etc